### ISSUE TITLE
Topics on iOS

### ISSUE THREAD

**Issue by ReallySmallSoftware** [chemerisuk/cordova-plugin-firebase-messaging#77]
I have implemented this plugin on Android and it works really well. I am struggling with iOS however.

I am trying to work with topics and didReceiveRegistrationToken() never seems to fire. I constantly get 'The subscription operation is suspended because you don't have a token. The operation will resume once you get an FCM token.' in the logs and never seem to get a token - I am calling getToken() and *do* get a token back from that and I am trying to subscribe after I perform that operation.

If I send a notification using the token it does arrive on the iOS device, so something is at least partly working. The topics never manage to subscribe however.

I'm partly after some reassurance that topics on iOS have been seen to be working as I am using a number of other Firebase plugins and although I don't believe there is any interaction I'm not ruling it out.

**Comment by ReallySmallSoftware** [comment]
I've done some more digging and have been able to get this to work. I'll try and detail my findings below.

This is tested on iOS 10 - maybe a later version would have behaved differently.

It all comes down to timing. I basically had to add the following into didFinishLaunchingWithOptions

```
    if(![FIRApp defaultApp]) {
        [FIRApp configure];
    }
```

Such that this now looks like this:

```
- (BOOL)identity_application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    
    if(![FIRApp defaultApp]) {
        [FIRApp configure];
    }
    
    [FIRMessaging messaging].delegate = self;
    [UNUserNotificationCenter currentNotificationCenter].delegate = self;

    if (launchOptions) {
.
.
.
```

This is because the plugin initialise was happening too late and the setting of the delegate in the above code caused the following code in FIRMessaging.m to be triggered before the messaging instance was set:

```
  FIRMessaging *messaging = (FIRMessaging *)instance;

  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    [messaging start];
  });
```

The dispatch_once is indeed dispatching once, but messaging is nil, so start never gets called and things don't get set up.

This is what I found - YMMV, but I can stop pulling my hair out now!

On the whole this is a great plugin and has enabled me to replace phonegap-push and is a better fit with my firebase plugins IMOO.

**Comment by chemerisuk** [comment]
@ReallySmallSoftware this is a great finding indeed. I've found that sometimes `getToken` is not resolved with instanceId. It could be the cause. The [documentation](https://firebase.google.com/docs/cloud-messaging/ios/client) says initialization code should be in the `application:didFinishLaunchingWithOptions` method.

Would you mind to submit a PR with the changes?

**Comment by ReallySmallSoftware** [comment]
@chemerisuk Sure - will try and get you something soon

**Comment by chemerisuk** [comment]
The fix publish in v2.0.1
